dynamic_programming/unique_paths.cpp

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file
+ * @brief Implementation of Unique Paths problem using Dynamic Programming.
+ * @details
+ * A robot is located at the top-left corner of an m × n grid.
+ * The robot can move either down or right at any point in time.
+ * This program computes the total number of unique paths to reach
+ * the bottom-right corner.
+ *
+ * Approaches:
+ * - **Memoization (Top-Down)**: Recursively explores solutions while
+ *   storing intermediate results in a cache (`memoization_table`) to avoid redundant
+ *   computation. Typically more intuitive and easy to write, but
+ *   relies on recursion and has associated call-stack overhead.
+ *
+ * - **Tabulation (Bottom-Up)**: Iteratively builds the solution using
+ *   a DP table, starting from the base cases and filling up the table.
+ *   Offers consistent performance without recursion overhead and
+ *   is generally more space-efficient when optimized.
+ *
+ * @see https://leetcode.com/problems/unique-paths/
+ */
+
+#include <iostream>
+#include <vector>
+#include <cassert>
+
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+
+/**
+ * @class UniquePathsSolver
+ * @brief Solves the Unique Paths problem using both memoization and tabulation.
+ */
+class UniquePathsSolver {
+   private:
+    std::vector<std::vector<int>> memoization_table;  ///< Memoization table
+    int m, n;
+
+    /**
+     * @brief Bottom-up Tabulation solution.
+     * @return int Number of unique paths from (0, 0) to (m-1, n-1)
+     */
+    int solveTabulation() {
+        std::vector<std::vector<int>> table(m, std::vector<int>(n, 0));
+
+        for (int i = 0; i < m; i++) table[i][n - 1] = 1;  ///< last column
+        for (int j = 0; j < n; j++) table[m - 1][j] = 1;  ///< last row
+
+        for (int i = m - 2; i >= 0; i--) {
+            for (int j = n - 2; j >= 0; j--) {
+                table[i][j] = table[i + 1][j] + table[i][j + 1];
+            }
+        }
+        return table[0][0];
+    }
+
+   public:
+    /**
+     * @brief Constructor initializes dimensions and memoization table
+     */
+    UniquePathsSolver(int rows, int cols) : m(rows), n(cols) {
```

Comment:
Can rows and columns be negative? I don't think so. Use fixed width unsigned integers like `size_t` or `uint32_t`.